proc/rename.go

```diff
@@ -0,0 +1,113 @@
+package proc
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/brimsec/zq/ast"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+)
+
+// Rename renames one or more fields in a record. A field can only be
+// renamed within its own record. For example id.orig_h can be
+// renamed to id.src, but it cannot be renamed to src. Renames are
+// applied left to right; each rename observes the effect of all
+// renames that preceded it.
+type Rename struct {
+	Base
+	fieldnames []string
+	targets    []string
+	typeMap    map[int]*zng.TypeRecord
```

Comment:
Is there a real need for type IDs here? If not, just make this `map[zng.Type]*zng.TypeRecord`.

Reply:
Could you explain the rationale please?

Reply:
Using the types themselves is, I think, clearer than using their IDs. Specifically, the role played here by a `map[zng.Type]*zng.TypeRecord` is more obvious than that played by a `map[int]*zng.TypeRecord`.

Reply:
We have many places where there's a `map[int]something` where the key is the return of `zng.Type.ID()`. Would you suggest we have `Type.ID()` return something other than `int` & use that?

Reply:
I see. Seems reasonable though I think it should be `map[zng.TypeRecord]*zng.TypeRecord`. 

Though if this is all about type descriptiveness, then an alternative would be to declare `type TypeID int` in `zng/type.go` (as we have done elsewhere) and switch to that type in place of ints, including using a  `map[zng.TypeID]*zng.TypeRecord` right here.

Reply:
> We have many places where there's a map[int]something where the key is the return of zng.Type.ID(). Would you suggest we have Type.ID() return something other than int & use that?

@alfred-landrum: No, I'd instead suggest that we use `map[*zng.RecordType]something` or `map[zng.Type]something` instead of `map[int]something` whenever practical because I think they're clearer. (They're also marginally faster.)

Reply:
> Though if this is all about type descriptiveness, then an alternative would be to declare `type TypeID int` in `zng/type.go` (as we have done elsewhere) and switch to that type in place of ints, including using a `map[zng.TypeID]*zng.TypeRecord` right here.

@henridf: After some reflection, the ultimate issue for me here is my sense that type IDs shouldn't be part of the zng.Type interface. That's clearly a topic for another venue, and as @alfred-landrum pointed out, there's plenty of precedent for what you've done here, so I think you should leave it as is.

Sorry it took me so long to figure that out!